### ISSUE TITLE
Storage: Why are fields base64 encoded?

### ISSUE THREAD

**Issue by ccoenen** [hackmdio/hackmd#239]
I am currently looking at the database model, and Many of the fields are stored base64 encoded. Is there any particular reason for this? I find it a little bit odd, I haven't seen this approach in other tools, so far.

It certainly makes third party development a little harder.

**Comment by jackycute** [comment]
Hi @ccoenen,
It's not just base64 encoded, it's also compressed.
Even you decode them with base64, you can't read them.

So we use [LZString](https://github.com/pieroxy/lz-string) to compress the text and base64 for storing them.
It can make to ensure content integrity and not saving them in plaintext, more over reduce data size at same time.

I knew this might bring more difficulties to third party dev, maybe open some API will be better.
see #11 

We still can talk about this.

**Comment by ccoenen** [comment]
I think, a database system should already ensure content integrity. At least databases following the [ACID](https://en.wikipedia.org/wiki/ACID) principles. I'm not entirely sure about document-oriented or key-value oriented database types, but those are not used at the moment anyway.

I was just wondering, because this also has the side-effect of making full-text-search impractical.

**Comment by jackycute** [comment]
Hey, sorry for late reply.
Basically you're right, we need to make a compress layer on DB instead of alter original data.
And I found the same question as you about the fulltext search, but it might able to put replica to search engine for indexing only usage.

Anyway this might to be done, make all data in DB to be original - not compressed or base64.

PS: we need to inform our users to turn on long index when they use utf8mb4 (for storing emoji) as charset.

Thanks for heads up.

**Comment by ccoenen** [comment]
I spent a little time over the past five days to think about this. Personally, I would come up with this pro/contra list:

+ space (Not a problem in my case, probably interesting if you have a ton of users on the same machine)
+ might prevent encoding issues (not a problem I encounter often in other projects)
+ provides a tiny layer of obfuscation (i.e. a database admin will not accidentally read a very personal note by accident)
+ provides an extra integrity check (also not something I ever really worried about in database backed projects)
+ keeping it, because changing something that currently works may introduce bugs

- It's unusual (not a big deal, just has to be documented somewhere)
- Makes third-party development a little harder (third party directly on the same database is unusual)
- Full text search is a lot harder (this is a *big* one for me)

This may not be exhaustive, in fact if anyone has other pro/contra items, I'll happily add them! To me, things on the pro list are not terribly important, so my personal opinion would be in favor of plain text.

**Comment by rufuspollock** [comment]
My 2c: do compression in the DB layer not the application layer. Optimization like these add cognitive and code complexity and probably turn out to be a premature. I'd keep the app layer simple and if space and performance really is an issue you can probably introduce and tune at the DB layer.

**Comment by jackycute** [comment]
@rufuspollock I agreed.
This will ship and being a breaking change in 0.5.0.
Coming with a migration tool though.

**Comment by jackycute** [comment]
I found a problem that some text currently saving contains non-utf8 character (such as half emoji).
If we lose base64 encode, users won't able to write/save non-utf8 content.
I have no idea how to progress now.

**Comment by jackycute** [comment]
Also considering about malformed utf-8 character too.
Maybe https://github.com/mathiasbynens/utf8.js or https://github.com/mathiasbynens/wtf-8 would help.

**Comment by ccoenen** [comment]
if this is about the conversion tool, I'd suggest a preview run that checks if this is a problem at all, reports it (i.e. reports all cases where this will occur) and then stops before any data is actually altered (so the user doesn't get stuck between half of his data converted and half of it base64).

**Comment by jackycute** [comment]
@ccoenen It's not quite possible to know what will happen before write in.
So what about tell users to backup their database then to try?

What I'm currently doing is using partial data from server to test to migration.
It will only release when it 100% migrate the data and new server running smoothly.

**Comment by jackycute** [comment]
OK, so the postgresql won't accept any null byte in the text datatype.
Now we have two options:
1. keep the null byte and choose one encoding (see above) which will only apply on the user provide content.
2. drop the null byte and prevent user input them, no need extra encoding but need filter null byte out before saving to db.

How do you say? @ccoenen @rufuspollock

**Comment by jackycute** [comment]
Personally I would like to escape or remove null byte out.

Found some informations here:
https://github.com/sequelize/sequelize/issues/6485
https://www.owasp.org/index.php/Embedding_Null_Code

**Comment by ccoenen** [comment]
one could also migrate into a different field or table (following field names are made-up):

- Read from legacy compressed field `text`
- convert
- write to new field `text_a`
- when everything went fine, drop field `text`

all new code uses `text_a` (the places have to be touched anyway). All legacy could would have used `text`. In case of a conversion problem, no legacy data has been changed, you could just continue using a 0.4 version for a little while longer. Most importantly, you don't have half of your database in one state and one half in the other state. Also nice: one could pause and resume the conversion - all fields with empty `text_a` still need conversion.

The problem with this approach is, that the current database model ist very nice and the field names are very well fitting. Changing over to different field or table names would either need some kind of suffix to the name or using a synonym that might not be as fitting.

**Comment by ccoenen** [comment]
Regarding the null-byte: I haven't personally encountered that problem so far. Is this related to a specific language or emoji set or something? Or is there just the general possibility "there could be a null-byte somewhere`.

For the latter: drop and prevent is fine. At least i can't think of a useful application inside markdown.

**Comment by jackycute** [comment]
To any user who don't want to run the migration, stay at the old version.
Backup before doing the migration sounds acceptable to me.

The null byte issue is real thing I've encountered for these days when making the migration tool.
So as I mentioned as above, PostgreSQL can't take the text contains null type to text type column.

Drop the null byte won't only might affect the past revision and content, the length and diff patch which contains null byte might not be accurate.
The special unicode characters (not visible) might occur when diffing the emoji (because it's two unicode width) or 
or pasting some text with special characters (null byte), [read more here](https://mathiasbynens.be/notes/javascript-escapes)
But even the special unicode character occur, PostgreSQL which encode set to "utf8" still able to save it, only the except the null byte (I think it's because Security issue).

TL;DR
1. Keep in old version if you don't want to run migration, no difference for now.
2. It won't be able to input/paste null byte after 0.5.0, db will also remove them before saving. (still not affecting most use case)

**Comment by jackycute** [comment]
Ok, so I'll try to auto make backup and save logs.

**Comment by jackycute** [comment]
Fixed in 0.5.0
